embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/EmbraceImpl.kt

```diff
@@ -0,0 +1,1072 @@
+     */
+    fun endMoment(name: String, identifier: String?, properties: Map<String, Any>?) {
+        if (checkSdkStartedAndLogPublicApiUsage("end_moment")) {
+            eventService?.endEvent(name, identifier, normalizeProperties(properties, logger))
+            onActivityReported()
+        }
+    }
+
+    /**
+     * Signals that the app has completed startup.
+     *
+     * @param properties properties to include as part of the startup moment
+     */
+    fun endAppStartup(properties: Map<String, Any>?) {
+        endMoment(EmbraceEventService.STARTUP_EVENT_NAME, null, properties)
+    }
+
+    fun getTraceIdHeader(): String {
+        if (configService != null && checkSdkStarted("get_trace_id_header", false)) {
+            return configService?.networkBehavior?.getTraceIdHeader() ?: NetworkBehavior.CONFIG_TRACE_ID_HEADER_DEFAULT_VALUE
+        }
+        return NetworkBehavior.CONFIG_TRACE_ID_HEADER_DEFAULT_VALUE
+    }
+
+    fun generateW3cTraceparent(): String = generateW3CTraceparent()
+
+    fun recordNetworkRequest(request: EmbraceNetworkRequest) {
+        if (checkSdkStartedAndLogPublicApiUsage("record_network_request")) {
+            logNetworkRequest(request)
+        }
+    }
+
+    private fun logNetworkRequest(request: EmbraceNetworkRequest) {
+        if (configService?.networkBehavior?.isUrlEnabled(request.url) == true) {
+            networkLoggingService?.logNetworkRequest(request)
+            onActivityReported()
+        }
+    }
+
+    fun logMessage(message: String, severity: Severity, properties: Map<String, Any>?) {
+        logMessage(
+            fromSeverity(severity),
+            message,
+            properties,
+            null,
+            null,
+            LogExceptionType.NONE,
+            null,
+            null
+        )
+    }
+
+    fun logException(throwable: Throwable, severity: Severity, properties: Map<String, Any>?, message: String?) {
+        val exceptionMessage = if (throwable.message != null) throwable.message else ""
+        logMessage(
+            fromSeverity(severity),
+            (
+                message
+                    ?: exceptionMessage
+                ) ?: "",
+            properties,
+            throwable.getSafeStackTrace(),
+            null,
+            LogExceptionType.HANDLED,
+            null,
+            null,
+            throwable.javaClass.simpleName,
+            exceptionMessage
+        )
+    }
+
+    fun logCustomStacktrace(
+        stacktraceElements: Array<StackTraceElement>,
+        severity: Severity,
+        properties: Map<String, Any>?,
+        message: String?
+    ) {
+        logMessage(
+            fromSeverity(severity),
+            message
+                ?: "",
+            properties,
+            stacktraceElements,
+            null,
+            LogExceptionType.HANDLED,
+            null,
+            null,
+            null,
+            message
+        )
+    }
+
+    @JvmOverloads
```

Comment:
I hope we can get rid of these in a subsequent change because our code calling into this shouldn't be Java anymore